File: src/api/onenote.ts
import { MicrosoftAuthManager } from "./microsoft-auth";

const LOG = "[Kindle Scribe] OneNote:";

/** OneNote notebook from the Graph API. */
export interface OneNoteNotebook {
  id: string;
  displayName: string;
  lastModifiedDateTime: string;
  createdDateTime: string;
  isDefault: boolean;
  sectionsUrl: string;
}

/** OneNote section within a notebook. */
export interface OneNoteSection {
  id: string;
  displayName: string;
  lastModifiedDateTime: string;
  pagesUrl: string;
}

/** OneNote page metadata. */
export interface OneNotePage {
  id: string;
  title: string;
  createdDateTime: string;
  lastModifiedDateTime: string;
  contentUrl: string;
  level: number;
  order: number;
}

/**
 * Client for Microsoft Graph OneNote API.
 * Fetches notebooks, sections, pages, and page content.
 */
export class OneNoteClient {
  constructor(private auth: MicrosoftAuthManager) {}

  /** List all OneNote notebooks for the user. */
  async listNotebooks(): Promise<OneNoteNotebook[]> {
    console.log(`${LOG} Listing notebooks...`);
    const data = await this.auth.graphRequest("/me/onenote/notebooks");
    console.log(`${LOG} Found ${data.value?.length || 0} notebooks`);
    return data.value || [];
  }

  /** List sections in a notebook. */
  async listSections(notebookId: string): Promise<OneNoteSection[]> {
    console.log(`${LOG} Listing sections for notebook ${notebookId}...`);
    const data = await this.auth.graphRequest(
      `/me/onenote/notebooks/${notebookId}/sections`
    );
    return data.value || [];
  }

  /** List pages in a section. */
  async listPages(sectionId: string): Promise<OneNotePage[]> {
    console.log(`${LOG} Listing pages for section ${sectionId}...`);
    const data = await this.auth.graphRequest(
      `/me/onenote/sections/${sectionId}/pages`
    );
    return data.value || [];
  }

  /** Get the HTML content of a page. */
  async getPageContent(pageId: string): Promise<string> {
    console.log(`${LOG} Fetching content for page ${pageId}...`);
    const html = await this.auth.graphRequestHtml(
      `/me/onenote/pages/${pageId}/content`
    );
    return html;
  }

  /**
   * Get all pages from a specific notebook, organized by section.
   */
  async getAllPagesFromNotebook(
    notebookId: string
  ): Promise<Array<{ section: OneNoteSection; pages: OneNotePage[] }>> {
    const sections = await this.listSections(notebookId);
    const results: Array<{ section: OneNoteSection; pages: OneNotePage[] }> = [];

    for (const section of sections) {
      const pages = await this.listPages(section.id);
      results.push({ section, pages });
    }

    return results;
  }
}

/**
 * Convert OneNote HTML content to clean Markdown.
 * OneNote returns HTML with specific patterns we can handle.
 */
export function htmlToMarkdown(html: string): string {
  let md = html;

  // Remove XML declaration and head section
  md = md.replace(/<\?xml[^>]*\?>/gi, "");
  md = md.replace(/<head[^>]*>[\s\S]*?<\/head>/gi, "");
  md = md.replace(/<\/?html[^>]*>/gi, "");
  md = md.replace(/<\/?body[^>]*>/gi, "");

  // Convert headings
  md = md.replace(/<h1[^>]*>([\s\S]*?)<\/h1>/gi, "# $1\n\n");
  md = md.replace(/<h2[^>]*>([\s\S]*?)<\/h2>/gi, "## $1\n\n");
  md = md.replace(/<h3[^>]*>([\s\S]*?)<\/h3>/gi, "### $1\n\n");
  md = md.replace(/<h4[^>]*>([\s\S]*?)<\/h4>/gi, "#### $1\n\n");

  // Convert bold and italic
  md = md.replace(/<(strong|b)[^>]*>([\s\S]*?)<\/(strong|b)>/gi, "**$2**");
  md = md.replace(/<(em|i)[^>]*>([\s\S]*?)<\/(em|i)>/gi, "*$2*");
  md = md.replace(/<u[^>]*>([\s\S]*?)<\/u>/gi, "$1");

  // Convert lists
  md = md.replace(/<li[^>]*>([\s\S]*?)<\/li>/gi, "- $1\n");
  md = md.replace(/<\/?[ou]l[^>]*>/gi, "\n");

  // Convert links
  md = md.replace(/<a[^>]*href="([^"]*)"[^>]*>([\s\S]*?)<\/a>/gi, "[$2]($1)");

  // Convert line breaks and paragraphs
  md = md.replace(/<br\s*\/?>/gi, "\n");
  md = md.replace(/<p[^>]*>([\s\S]*?)<\/p>/gi, "$1\n\n");
  md = md.replace(/<div[^>]*>([\s\S]*?)<\/div>/gi, "$1\n");

  // Convert images (OneNote embeds)
  md = md.replace(/<img[^>]*src="([^"]*)"[^>]*alt="([^"]*)"[^>]*\/?>/gi, "![$2]($1)");
  md = md.replace(/<img[^>]*src="([^"]*)"[^>]*\/?>/gi, "![]($1)");

  // Remove remaining HTML tags
  md = md.replace(/<[^>]+>/g, "");

  // Decode HTML entities
  md = md.replace(/&amp;/g, "&");
  md = md.replace(/&lt;/g, "<");
  md = md.replace(/&gt;/g, ">");
  md = md.replace(/&quot;/g, '"');
  md = md.replace(/&#39;/g, "'");
  md = md.replace(/&nbsp;/g, " ");

  // Clean up whitespace
  md = md.replace(/\n{3,}/g, "\n\n");
  md = md.trim();

  return md;
}

File: kindle_sync/onenote.py
"""OneNote API client and HTML-to-Markdown converter."""

import json
import re
from urllib.request import Request, urlopen

from .config import MS_GRAPH_URL


class OneNoteClient:
    """Client for Microsoft Graph OneNote API."""

    def __init__(self, get_token):
        self._get_token = get_token

    def list_notebooks(self) -> list[dict]:
        """List all OneNote notebooks."""
        data = self._graph_get("/me/onenote/notebooks")
        return data.get("value", [])

    def list_sections(self, notebook_id: str) -> list[dict]:
        """List sections in a notebook."""
        data = self._graph_get(f"/me/onenote/notebooks/{notebook_id}/sections")
        return data.get("value", [])

    def list_pages(self, section_id: str) -> list[dict]:
        """List pages in a section."""
        data = self._graph_get(f"/me/onenote/sections/{section_id}/pages")
        return data.get("value", [])

    def get_page_content(self, page_id: str) -> str:
        """Get the HTML content of a page."""
        return self._graph_get_html(f"/me/onenote/pages/{page_id}/content")

    def _graph_get(self, endpoint: str) -> dict:
        """Make an authenticated GET request returning JSON."""
        url = f"{MS_GRAPH_URL}{endpoint}"
        req = Request(url)
        req.add_header("Authorization", f"Bearer {self._get_token()}")
        req.add_header("Content-Type", "application/json")

        with urlopen(req) as resp:
            return json.loads(resp.read())

    def _graph_get_html(self, endpoint: str) -> str:
        """Make an authenticated GET request returning HTML."""
        url = f"{MS_GRAPH_URL}{endpoint}"
        req = Request(url)
        req.add_header("Authorization", f"Bearer {self._get_token()}")
        req.add_header("Accept", "text/html")

        with urlopen(req) as resp:
            return resp.read().decode("utf-8")


def html_to_markdown(html: str) -> str:
    """Convert OneNote HTML content to clean Markdown."""
    md = html

    # Remove XML declaration and head section
    md = re.sub(r"<\?xml[^>]*\?>", "", md, flags=re.IGNORECASE)
    md = re.sub(r"<head[^>]*>[\s\S]*?</head>", "", md, flags=re.IGNORECASE)
    md = re.sub(r"</?html[^>]*>", "", md, flags=re.IGNORECASE)
    md = re.sub(r"</?body[^>]*>", "", md, flags=re.IGNORECASE)

    # Convert headings
    md = re.sub(r"<h1[^>]*>([\s\S]*?)</h1>", r"# \1\n\n", md, flags=re.IGNORECASE)
    md = re.sub(r"<h2[^>]*>([\s\S]*?)</h2>", r"## \1\n\n", md, flags=re.IGNORECASE)
    md = re.sub(r"<h3[^>]*>([\s\S]*?)</h3>", r"### \1\n\n", md, flags=re.IGNORECASE)
    md = re.sub(r"<h4[^>]*>([\s\S]*?)</h4>", r"#### \1\n\n", md, flags=re.IGNORECASE)

    # Convert bold and italic
    md = re.sub(r"<(strong|b)[^>]*>([\s\S]*?)</(strong|b)>", r"**\2**", md, flags=re.IGNORECASE)
    md = re.sub(r"<(em|i)[^>]*>([\s\S]*?)</(em|i)>", r"*\2*", md, flags=re.IGNORECASE)
    md = re.sub(r"<u[^>]*>([\s\S]*?)</u>", r"\1", md, flags=re.IGNORECASE)

    # Convert lists
    md = re.sub(r"<li[^>]*>([\s\S]*?)</li>", r"- \1\n", md, flags=re.IGNORECASE)
    md = re.sub(r"</?[ou]l[^>]*>", "\n", md, flags=re.IGNORECASE)

    # Convert links
    md = re.sub(r'<a[^>]*href="([^"]*)"[^>]*>([\s\S]*?)</a>', r"[\2](\1)", md, flags=re.IGNORECASE)

    # Convert line breaks and paragraphs
    md = re.sub(r"<br\s*/?>", "\n", md, flags=re.IGNORECASE)
    md = re.sub(r"<p[^>]*>([\s\S]*?)</p>", r"\1\n\n", md, flags=re.IGNORECASE)
    md = re.sub(r"<div[^>]*>([\s\S]*?)</div>", r"\1\n", md, flags=re.IGNORECASE)
    md = re.sub(r"<span[^>]*>([\s\S]*?)</span>", r"\1", md, flags=re.IGNORECASE)

    # Remove remaining HTML tags
    md = re.sub(r"<[^>]+>", "", md)

    # Decode HTML entities
    md = md.replace("&amp;", "&")
    md = md.replace("&lt;", "<")
    md = md.replace("&gt;", ">")
    md = md.replace("&quot;", '"')
    md = md.replace("&#39;", "'")
    md = md.replace("&nbsp;", " ")

    # Clean up whitespace
    md = re.sub(r"\n{3,}", "\n\n", md)
    md = md.strip()

    return md
